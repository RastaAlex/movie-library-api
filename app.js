import express from 'express';
import bodyParser from 'body-parser';
import cors from 'cors';
import helmet from 'helmet';
import rateLimit from 'express-rate-limit';
import dotenv from 'dotenv';
import movieRoutes from './src/routes/moviesRoutes.js';
import userRoutes from './src/routes/usersRoutes.js';
import { sequelize } from './src/config/database.js';

dotenv.config();
const app = express();

const limiter = rateLimit({
  windowMs: 15 * 60 * 1000,
  max: 100,
});

app.use(limiter);

app.use(cors());
app.use(helmet());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

app.use(userRoutes);
app.use(movieRoutes);

const PORT = process.env.PORT || 8000;

const startServer = async () => {
  try {
    await sequelize.sync();
    app.listen(PORT, () => {
      console.log(`Server is running on port ${PORT}`);
    });
  } catch (error) {
    console.error('Unable to connect to the database:', error);
  }
};

if (process.env.NODE_ENV !== 'test') {
  startServer();
}

export { app };
