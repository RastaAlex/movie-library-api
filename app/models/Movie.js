import { Sequelize, DataTypes, Model } from 'sequelize';
import { sequelize } from '../../config/database.js';

class Movie extends Model {}

Movie.init({
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true,
    unique: true,
  },
  title: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  releaseYear: {
    type: DataTypes.INTEGER,
    allowNull: false,
  },
  format: {
    type: DataTypes.ENUM('VHS', 'DVD', 'Blu-ray'),
    allowNull: false,
  },
  actors: {
    type: DataTypes.STRING,
    allowNull: false,
  },
}, {
  sequelize,
  modelName: 'Movie',
  timestamps: false,
});

export default Movie;
