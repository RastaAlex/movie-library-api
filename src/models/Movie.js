import { DataTypes, Model } from 'sequelize';
import { sequelize } from '../../config/database.js';

class Movie extends Model {}

Movie.init(
  {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
      unique: true,
    },
    title: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    year: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    format: {
      type: DataTypes.ENUM('VHS', 'DVD', 'Blu-ray'),
      allowNull: false,
    },
    actors: {
      type: DataTypes.STRING,
      allowNull: false,
      get() {
        return this.getDataValue('actors').split(';');
      },
      set(val) {
        this.setDataValue('actors', val.join(';'));
      },
    },
  },
  {
    sequelize,
    modelName: 'Movie',
    timestamps: false,
  },
);

export default Movie;
