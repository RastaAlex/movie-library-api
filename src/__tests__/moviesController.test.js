import supertest from 'supertest';
import { app } from '../../app.js';
import { sequelize } from '../../config/database.js';
import Movie from '../models/Movie.js';

jest.mock('../middleware/authMiddleware.js', () => ({
  __esModule: true,
  default: jest.fn((req, res, next) => {
    req.user = { id: 1 };
    next();
  }),
}));

let server;
let request;

beforeAll(async () => {
  await sequelize.sync({ force: true });
  server = app.listen(0);
  request = supertest(server);
});

afterAll(async () => {
  await sequelize.close();
  await server.close();
});

describe('Movies API', () => {
  let testMovie;

  beforeEach(async () => {
    testMovie = await Movie.create({
      title: 'Test Movie',
      year: 2023,
      format: 'DVD',
      actors: ['Test Actor1', 'Test Actor2'],
    });
  });

  afterEach(async () => {
    await Movie.destroy({ where: {}, truncate: true });
  });

  test('GET should return a list of movies', async () => {
    const response = await request.get('/api/v1/movies');
    expect(response.statusCode).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  test('GET should return a movie by id', async () => {
    const response = await request.get(`/api/v1/movies/${testMovie.id}`);
    expect(response.statusCode).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  test('POST should create a new movie', async () => {
    const newMovie = {
      title: 'New Movie',
      year: 2023,
      format: 'DVD',
      actors: ['New Actor1', 'New Actor2'],
    };

    const response = await request.post('/api/v1/movies').send(newMovie);
    expect(response.statusCode).toBe(201);
    expect(response.body).toMatchSnapshot();
  });

  test('PATCH should update a movie', async () => {
    const updatedMovie = {
      title: 'Updated Movie',
      year: 2023,
      format: 'Blu-Ray',
      actors: ['Updated Actor1', 'Updated Actor2'],
    };

    const response = await request.patch(`/api/v1/movies/${testMovie.id}`).send(updatedMovie);
    expect(response.statusCode).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  test('DELETE should delete a movie', async () => {
    const response = await request.delete(`/api/v1/movies/${testMovie.id}`);
    expect(response.statusCode).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  test('POST should import movies from a file', async () => {
    const filePath = 'src/__tests__/testFile.txt';
    const response = await request.post('/api/v1/movies/import').attach('file', filePath);
    expect(response.statusCode).toBe(200);
    expect(response.body).toMatchSnapshot();
  });
});
