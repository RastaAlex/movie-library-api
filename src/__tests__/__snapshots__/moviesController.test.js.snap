// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies API DELETE /movies/:id should delete a movie 1`] = `
{
  "status": 1,
}
`;

exports[`Movies API DELETE should delete a movie 1`] = `
{
  "status": 1,
}
`;

exports[`Movies API GET should return a list of movies 1`] = `
[
  {
    "actors": [
      "Test Actor1",
      "Test Actor2",
    ],
    "format": "DVD",
    "id": 1,
    "title": "Test Movie",
    "year": 2023,
  },
]
`;

exports[`Movies API GET should return a movie by id 1`] = `
{
  "data": {
    "actors": [
      "Test Actor1",
      "Test Actor2",
    ],
    "format": "DVD",
    "id": 2,
    "title": "Test Movie",
    "year": 2023,
  },
  "status": 1,
}
`;

exports[`Movies API PATCH /movies/:id should update a movie 1`] = `
{
  "data": {
    "actors": [
      "Updated Actor1",
      "Updated Actor2",
    ],
    "format": "Blu-Ray",
    "id": 5,
    "title": "Updated Movie",
    "year": 2023,
  },
  "status": 1,
}
`;

exports[`Movies API PATCH should update a movie 1`] = `
{
  "data": {
    "actors": [
      "Updated Actor1",
      "Updated Actor2",
    ],
    "format": "Blu-Ray",
    "id": 5,
    "title": "Updated Movie",
    "year": 2023,
  },
  "status": 1,
}
`;

exports[`Movies API POST /movies should create a new movie 1`] = `
{
  "data": {
    "actors": [
      "New Actor1",
      "New Actor2",
    ],
    "format": "DVD",
    "id": 4,
    "title": "New Movie",
    "year": 2023,
  },
  "status": 1,
}
`;

exports[`Movies API POST should create a new movie 1`] = `
{
  "data": {
    "actors": [
      "New Actor1",
      "New Actor2",
    ],
    "format": "DVD",
    "id": 4,
    "title": "New Movie",
    "year": 2023,
  },
  "status": 1,
}
`;

exports[`Movies API POST should import movies from a file 1`] = `
{
  "data": [
    {
      "actors": [
        "Mel Brooks",
        "Clevon Little",
        "Harvey Korman",
        "Gene Wilder",
        "Slim Pickens",
        "Madeline Kahn",
      ],
      "format": "VHS",
      "id": 8,
      "title": "Blazing Saddles",
      "year": 1974,
    },
    {
      "actors": [
        "Humphrey Bogart",
        "Ingrid Bergman",
        "Claude Rains",
        "Peter Lorre",
      ],
      "format": "DVD",
      "id": 9,
      "title": "Casablanca",
      "year": 1942,
    },
    {
      "actors": [
        "Audrey Hepburn",
        "Cary Grant",
        "Walter Matthau",
        "James Coburn",
        "George Kennedy",
      ],
      "format": "DVD",
      "id": 10,
      "title": "Charade",
      "year": 1953,
    },
    {
      "actors": [
        "Paul Newman",
        "George Kennedy",
        "Strother Martin",
      ],
      "format": "VHS",
      "id": 11,
      "title": "Cool Hand Luke",
      "year": 1967,
    },
    {
      "actors": [
        "Paul Newman",
        "Robert Redford",
        "Katherine Ross",
      ],
      "format": "VHS",
      "id": 12,
      "title": "Butch Cassidy and the Sundance Kid",
      "year": 1969,
    },
    {
      "actors": [
        "Robert Redford",
        "Paul Newman",
        "Robert Shaw",
        "Charles Durning",
      ],
      "format": "DVD",
      "id": 13,
      "title": "The Sting",
      "year": 1973,
    },
    {
      "actors": [
        "Jim Henson",
        "Frank Oz",
        "Dave Geolz",
        "Mel Brooks",
        "James Coburn",
        "Charles Durning",
        "Austin Pendleton",
      ],
      "format": "DVD",
      "id": 14,
      "title": "The Muppet Movie",
      "year": 1979,
    },
    {
      "actors": [
        "John Travolta",
        "Danny DeVito",
        "Renne Russo",
        "Gene Hackman",
        "Dennis Farina",
      ],
      "format": "DVD",
      "id": 15,
      "title": "Get Shorty ",
      "year": 1995,
    },
    {
      "actors": [
        "Joe Pesci",
        "Marrisa Tomei",
        "Fred Gwynne",
        "Austin Pendleton",
        "Lane Smith",
        "Ralph Macchio",
      ],
      "format": "DVD",
      "id": 16,
      "title": "My Cousin Vinny",
      "year": 1992,
    },
    {
      "actors": [
        "Russell Crowe",
        "Joaquin Phoenix",
        "Connie Nielson",
      ],
      "format": "Blu-Ray",
      "id": 17,
      "title": "Gladiator",
      "year": 2000,
    },
    {
      "actors": [
        "Harrison Ford",
        "Mark Hamill",
        "Carrie Fisher",
        "Alec Guinness",
        "James Earl Jones",
      ],
      "format": "Blu-Ray",
      "id": 18,
      "title": "Star Wars",
      "year": 1977,
    },
    {
      "actors": [
        "Harrison Ford",
        "Karen Allen",
      ],
      "format": "DVD",
      "id": 19,
      "title": "Raiders of the Lost Ark",
      "year": 1981,
    },
    {
      "actors": [
        "Nathan Fillion",
        "Alan Tudyk",
        "Adam Baldwin",
        "Ron Glass",
        "Jewel Staite",
        "Gina Torres",
        "Morena Baccarin",
        "Sean Maher",
        "Summer Glau",
        "Chiwetel Ejiofor",
      ],
      "format": "Blu-Ray",
      "id": 20,
      "title": "Serenity",
      "year": 2005,
    },
    {
      "actors": [
        "Gene Hackman",
        "Barbara Hershey",
        "Dennis Hopper",
      ],
      "format": "VHS",
      "id": 21,
      "title": "Hooisers",
      "year": 1986,
    },
    {
      "actors": [
        "Matthew Broderick",
        "Ally Sheedy",
        "Dabney Coleman",
        "John Wood",
        "Barry Corbin",
      ],
      "format": "VHS",
      "id": 22,
      "title": "WarGames",
      "year": 1983,
    },
    {
      "actors": [
        "Bill Pullman",
        "John Candy",
        "Mel Brooks",
        "Rick Moranis",
        "Daphne Zuniga",
        "Joan Rivers",
      ],
      "format": "DVD",
      "id": 23,
      "title": "Spaceballs",
      "year": 1987,
    },
    {
      "actors": [
        "Gene Wilder",
        "Kenneth Mars",
        "Terri Garr",
        "Gene Hackman",
        "Peter Boyle",
      ],
      "format": "VHS",
      "id": 24,
      "title": "Young Frankenstein",
      "year": 1974,
    },
    {
      "actors": [
        "Val Kilmer",
        "Gabe Jarret",
        "Michelle Meyrink",
        "William Atherton",
      ],
      "format": "VHS",
      "id": 25,
      "title": "Real Genius",
      "year": 1985,
    },
    {
      "actors": [
        "Tom Cruise",
        "Kelly McGillis",
        "Val Kilmer",
        "Anthony Edwards",
        "Tom Skerritt",
      ],
      "format": "DVD",
      "id": 26,
      "title": "Top Gun",
      "year": 1986,
    },
    {
      "actors": [
        "Donald Sutherland",
        "Elliot Gould",
        "Tom Skerritt",
        "Sally Kellerman",
        "Robert Duvall",
      ],
      "format": "DVD",
      "id": 27,
      "title": "MASH",
      "year": 1970,
    },
    {
      "actors": [
        "Carl Reiner",
        "Eva Marie Saint",
        "Alan Arkin",
        "Brian Keith",
      ],
      "format": "VHS",
      "id": 28,
      "title": "The Russians Are Coming, The Russians Are Coming",
      "year": 1966,
    },
    {
      "actors": [
        "Roy Scheider",
        "Robert Shaw",
        "Richard Dreyfuss",
        "Lorraine Gary ",
      ],
      "format": "DVD",
      "id": 29,
      "title": "Jaws",
      "year": 1975,
    },
    {
      "actors": [
        "Keir Dullea",
        "Gary Lockwood",
        "William Sylvester",
        "Douglas Rain",
      ],
      "format": "DVD",
      "id": 30,
      "title": "2001",
      "year": 1968,
    },
    {
      "actors": [
        "James Stewart",
        "Josephine Hull",
        "Peggy Dow",
        "Charles Drake",
      ],
      "format": "DVD",
      "id": 31,
      "title": "Harvey",
      "year": 1950,
    },
    {
      "actors": [
        "Seth Rogen",
        "Katherine Heigl",
        "Paul Rudd",
        "Leslie Mann",
      ],
      "format": "Blu-Ray",
      "id": 32,
      "title": "Knocked Up",
      "year": 2007,
    },
  ],
  "meta": {
    "imported": 25,
    "total": 25,
  },
  "status": 1,
}
`;
